### DISCUSSION TITLE
StringBuilders magic for very large strings

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#108] [General]
# StringBuilders magic for very large strings

The StringBuilder class is used to create mutable sequences of characters. Strings are immutable, so if you need to perform multiple operations on a string, it is better to use a StringBuilder instead of a string. This is especially useful when you need to concatenate a large number of strings. But there is more magic to it, especially when we go BIG!

https://steven-giesel.com/blogPost/c0144fea-0ff8-4638-be01-b62b48ef7188

**Comment by Patrick-3000** [comment]
Very nicely explained! So just that I get it right: the internal implementation of the `StringBuilder` class is responsible for the fact that `StringBuilder` instances dont live on the LOH, and this is a good thing because they can now be compacted on the "normal" heap? 
What I think I missed: where does the big difference (factor of 2) in memory allocation between `StringBuilder()` and `List()` come from?

**Reply by linkdotnet** [reply]
Basically, yes. The implementation detail that there is no single object that can grow bigger than 85'000 bytes is the reason that `StringBuilder` doesn't end up on the LOH.

The Large Object Heap can also be compacted, but you have to specifically ask for that:
```csharp
// Don't just "do" that. That is something you want to benchmark for your scenario
// Game-Engines are potentially benefiting from this
GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
GC.Collect();
```

Regarding the Memory of `List` and `StringBuilder` I did update the original blog post (see the last section). I hope that answers your question. If not let me know.

Grüsse
Steven

**Reply by Patrick-3000** [reply]
Now it is clear for me, thank you very much!
